sklearn/metrics/classification.py

```diff
@@ -1597,36 +1602,54 @@ def log_loss(y_true, y_pred, eps=1e-15, normalize=True, sample_weight=None):
     The logarithm used is the natural logarithm (base-e).
     """
     lb = LabelBinarizer()
-    T = lb.fit_transform(y_true)
-    if T.shape[1] == 1:
-        T = np.append(1 - T, T, axis=1)
 
+    if labels is not None:
+        lb.fit(labels)
+    else:
+        lb.fit(y_true)
+
+    if len(lb.classes_) == 1:
+        if labels is None:
+            raise ValueError('y_true has only one label. Please provide '
+                             'the true labels explicitly through the labels '
+                             'argument.')
+        else:
+            raise ValueError('labels should have more than one unique label.')
+
+    transformed_labels = lb.transform(y_true)
+
+    if transformed_labels.shape[1] == 1:
+        transformed_labels = np.append(1 - transformed_labels,
+                                       transformed_labels, axis=1)
     y_pred = check_array(y_pred, ensure_2d=False)
-    # Clipping
-    Y = np.clip(y_pred, eps, 1 - eps)
 
-    # This happens in cases when elements in y_pred have type "str".
-    if not isinstance(Y, np.ndarray):
-        raise ValueError("y_pred should be an array of floats.")
+    # Verify the dtype of y_pred is float.
+    if y_pred.dtype != float:
```

Comment:
yeah, i thought that would be a gotcha as well. but I ran the following and it made me think it was fine:

```
In [1]: import numpy as np

In [2]: l = np.array([.1,.2,.3,.4])

In [3]: l.dtype
Out[3]: dtype('float64')

In [4]: l.dtype == float
Out[4]: True

In [5]: l.dtype != float
Out[5]: False
```

looking at the failing tests, the `y_pred`s being passed in are of type `bool`, so doesn't seem related to that.